### PR TITLE
[NEW] Feature/livechat translation

### LINKED ISSUE #8105: [Feature Request] Translations for LiveChat
Rocket.Chat Version: 0.58.2

For any business that operates in more than one language, it is a requirement that the LiveChat plugin be the same language *as the host-page* that would be independent from Rocketchat's server language config.

Things to have translations for:
## User defined
- [x] Title
- [x] Unavailable message
- [x] Offline title
- [x] Offline message
- [x] Success message
- [ ] Custom fields labels

## RocketChat constants
- [X] Offline email form placeholders and buttons
- [X] "Type your message" placeholder
- [ ] The small option menu and its close chat button

### PR DESCRIPTION
**Pull request by icosamuel**
@RocketChat/core 
Closes #8105

# Translation of the Livechat #
In a business that operates in multiple languages, having the Livechat plugin be in the same language as the host-page is absolutely necessary. So I will guide you in my humble journey to make this happen.   
Many thanks to @sampaiodiego for helping and reviewing me throughout the process

## Configuration ##
 The new dropdown in the configurations represents the current language of the appearance settings. This is where you can add translations for your setup. Be sure to have at least an english version completed so the system has something to fallback on if a translation is missing or is incomplete.  

screenshot from 2017-12-07 10-30-40  

The settings are saved in the new Model `RocketChat.models.LivechatTexts` and Mongo collection `rocketchat_livechat_texts`. There are a new Meteor method `livechat:saveAppearanceTexts` to write and a publication `livechat:appearancetexts` to read.
``` javascript
{
    "_id" : "Lb9cF8EJY5diTyB7u",
    "identifier" : "Livechat_title",
    "lang" : "en",
    "text" : "english title"
},
{
    "_id" : "4zxkkGLoRHmuZFFCk",
    "identifier" : "Livechat_title",
    "lang" : "fr",
    "text" : "titre francais"
}
```

## On the client side ##

You can send your **host page's language** with this new API function (which I added in the demo)
```javascript
RocketChat(function() {
    this.setLanguage('pt-BR');
});
```
The setting is first stored in `Livechat.language` and then the two following mechanisms take place.

### User defined ###
The hook calls a new Meteor method `getTextsTranslation` in order to gather the **user defined** labels and texts. It then updates all the appropriate labels and texts used by Livechat.  
These information are also initially gathered by `getInitialData` before the API hook is triggered, but they will be loaded in your **server's language**. This is also the default behaviour if you don't use the API.

### Constants ###
Constants in the Livechat all use TAPi18n, so `TAPi18n.setLanguage(result.language);` takes care of that part. The only change is that it is now controlled by the host-page language if availiable, instead of the server's language.  
Note that the issue #8312 is blocking some minor parts from being properly translated.

### CODE REVIEW

File: packages/rocketchat-livechat/client/views/app/livechatAppearance.html [hunk 1]
```diff
@@ -5,6 +5,18 @@
 
 			<form class="rocket-form">
 				<fieldset>
+					<legend>Langue des texts</legend>
```
Comment (icosamuel) [inline packages/rocketchat-livechat/client/views/app/livechatAppearance.html]: Add a new label through i18n

File: packages/rocketchat-livechat/server/lib/Livechat.js [hunk 2]
```diff
@@ -252,6 +252,8 @@ RocketChat.Livechat = {
 			settings[setting._id] = setting.value;
 		});
 
+		// TODO: overwriter les configs qui existent dans la langue demandée
```
Comment (icosamuel) [inline packages/rocketchat-livechat/server/lib/Livechat.js]: // TODO: overwriter les configs qui existent dans la langue demandée

File: packages/rocketchat-livechat/server/methods/getInitialData.js [hunk 3]
```diff
@@ -34,6 +37,7 @@ Meteor.methods({
 			info.room = room[0];
 		}
 
+		// TODO: utiliser le language somehow pour loader les configs apporpriées
```
Comment (icosamuel) [inline packages/rocketchat-livechat/server/methods/getInitialData.js]: // TODO: utiliser le language somehow pour loader les configs apporpriées

File: packages/rocketchat-livechat/server/methods/saveAppearanceTexts.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+		const valid = settings.every((setting) => {
+			return validSettings.indexOf(setting._id) !== -1;
+		});
+
+		if (!valid) {
+			throw new Meteor.Error('invalid-setting');
+		}
+
+		settings.forEach((setting) => {
+			RocketChat.settings.updateById(setting._id, setting.value);
```
Comment (icosamuel) [inline packages/rocketchat-livechat/server/methods/saveAppearanceTexts.js]: @sampaiodiego How would I go about saving and fetching these new text settings?
As you recommended, the new object structure is 
```const settingTexts = [
	{
		identifier: 'Livechat_title',
		lang: instance.chosenLanguage,
		text: _.trim(instance.title.get())
	},
	[...]
];
```
Do I need to create a new model? Is there a similar part of the software I could base it on?

File: packages/rocketchat-livechat/client/views/app/livechatAppearance.js [hunk 5]
```diff
@@ -161,7 +182,7 @@ Template.livechatAppearance.onCreated(function() {
 	this.offlineSuccessMessage = new ReactiveVar(null);
 	this.titleOffline = new ReactiveVar(null);
 	this.colorOffline = new ReactiveVar(null);
-	this.offlineEmail = new ReactiveVar(null);
+	this.emailOffline = new ReactiveVar(null);
```
Comment (graywolf336) [inline packages/rocketchat-livechat/client/views/app/livechatAppearance.js]: Curious why this change? As `offlineEmail` means that the email to use for offline while `emailOffline` sounds like it should be a boolean or something stating of whether to email them offline or not..

Comment (icosamuel) [reply packages/rocketchat-livechat/client/views/app/livechatAppearance.js]: This seemed to fix a console error log that I don't remember on top of my head. Happened when the content of the offline email field changed.

You have a point about semantics. it sounds like a boolean, but all other variables are this way as well. `titleOffline`, `colorOffline`...

File: packages/rocketchat-lib/server/functions/livechatTexts.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+RocketChat.LivechatTexts = {};
+
+RocketChat.LivechatTexts.updateByNameAndLang = function(identifier, lang, value) {
+	if (identifier == null || value == null) {
+		return false;
+	}
+	lang = lang || 'en';
+
+	return RocketChat.models.LivechatTexts.upsertByNameAndLang(identifier, lang, value);
+};
```
[the remark this answers is not in the document]
Comment (icosamuel) [reply packages/rocketchat-lib/server/functions/livechatTexts.js]: What strategy do you use to enforce lint rules at pre-commit stage? It would save me so much time!

File: packages/rocketchat-livechat/app/client/lib/_livechat.js [hunk 7]
```diff
@@ -71,6 +72,10 @@ this.Livechat = new (class Livechat {
 	get title() {
 		return this._title.get();
 	}
+	get language() {
+		console.log(this._language);
```
Comment (icosamuel) [inline packages/rocketchat-livechat/app/client/lib/_livechat.js]: to remove

File: packages/rocketchat-livechat/app/client/lib/_livechat.js [hunk 8]
```diff
@@ -114,6 +119,10 @@ this.Livechat = new (class Livechat {
 	set title(value) {
 		this._title.set(value);
 	}
+	set language(value) {
+		console.log(this._language, value);
```
Comment (icosamuel) [inline packages/rocketchat-livechat/app/client/lib/_livechat.js]: to remove

File: packages/rocketchat-livechat/app/client/lib/_livechat.js [hunk 9]
```diff
@@ -2,6 +2,7 @@ this.Livechat = new (class Livechat {
 	constructor() {
 		this._online = new ReactiveVar(null);
 
+		this._language = new ReactiveVar();
```
Comment (icosamuel) [inline packages/rocketchat-livechat/app/client/lib/_livechat.js]: @sampaiodiego Since this is set asynchronously by the new livechat API function, it is not set when we first call GetInitialData. How can I best solve this issue?

Comment (icosamuel) [reply packages/rocketchat-livechat/app/client/lib/_livechat.js]: Unless you explicitly advise otherwise, I'm gonna fetch the translations with a new meteor function that will be called whenever the `Livechat.language` changes.

File: packages/rocketchat-lib/server/functions/livechatTexts.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
```
[the remark this answers is not in the document]
Comment (icosamuel) [reply packages/rocketchat-lib/server/functions/livechatTexts.js]: This might actually be a bot bug, am I wrong?

[the remark this answers is not in the document]
Comment (sampaiodiego) [reply packages/rocketchat-lib/server/functions/livechatTexts.js]: looks like it is =)

**Comment by icosamuel** [comment]
To complete this feature, there is a last piece of the puzzle that I can't figure out how to fit.
I need to find a way to control the translation language of static text like `'Close_menu'` or `'Are_you_sure_do_you_want_end_this_chat'`

How would I best solve this? I've tried multiple ways, but haven't stumbled upon something that works.

**Comment by sampaiodiego** [comment]
static translations would still be managed by `TAPi18n` package (as they are now).

are you experiencing the same as #8312?

**Comment by icosamuel** [comment]
Indeed! I made the livechat API control the Tapi18n language setting, but they still come out as english.

**Comment by sampaiodiego** [comment]
@icosamuel I'll test this so we can it merged to `0.61.0` ok? thx again

**Comment by icosamuel** [comment]
@sampaiodiego absolutely. In the meantime, I will update the description to reflect whats has been done. I've been running this for a while now, but remember that #8312 is still not fixed

**Comment by sampaiodiego** [comment]
I'm sorry I couldn't merge this before. Now the old/legacy livechat widget is gone so we can close this. 

If you have new suggestions or additions to the new livechat widget, please refer to https://github.com/RocketChat/Rocket.Chat.Livechat
